crates/ruff/src/rules/flake8_future_annotations/rules.rs

```diff
@@ -0,0 +1,109 @@
+use itertools::Itertools;
+use rustpython_parser::ast::{Alias, Expr, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_stdlib::typing::PEP_585_SUBSCRIPT_ELIGIBLE;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for missing `from __future__ import annotations` import if a type used in the
+/// module can be rewritten using PEP 563.
+///
+/// ## Why is this bad?
+/// Pairs well with pyupgrade with the --py37-plus flag or higher, since pyupgrade only
+/// replaces type annotations with the PEP 563 rules if `from __future__ import annotations`
+/// is present.
+///
+/// ## Example
+/// ```python
+/// import typing as t
+/// from typing import List
+///
+///
+/// def function(a_dict: t.Dict[str, t.Optional[int]]) -> None:
+///     a_list: List[str] = []
+///     a_list.append("hello")
+/// ```
+///
+/// To fix the lint error:
+/// ```python
+/// from __future__ import annotations
+///
+/// import typing as t
+/// from typing import List
+///
+///
+/// def function(a_dict: t.Dict[str, t.Optional[int]]) -> None:
+///     a_list: List[str] = []
+///     a_list.append("hello")
+/// ```
+///
+/// After running additional pyupgrade autofixes:
+/// ```python
+/// from __future__ import annotations
+///
+///
+/// def function(a_dict: dict[str, int | None]) -> None:
+///     a_list: list[str] = []
+///     a_list.append("hello")
+/// ```
+#[violation]
+pub struct MissingFutureAnnotationsWithImports {
+    pub names: Vec<String>,
+}
+
+impl Violation for MissingFutureAnnotationsWithImports {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let MissingFutureAnnotationsWithImports { names } = self;
+        let names = names.iter().map(|name| format!("`{name}`")).join(", ");
+        format!("Missing `from __future__ import annotations`, but imports: {names}")
+    }
+}
+
+/// FA100
+pub fn missing_future_annotations_from_typing_import(
+    checker: &mut Checker,
+    stmt: &Stmt,
+    module: &str,
+    names: &[Alias],
+) {
+    if checker.ctx.annotations_future_enabled {
+        return;
+    }
+
+    let names: Vec<String> = names
+        .iter()
+        .map(|name| name.node.name.as_str())
+        .filter(|alias| PEP_585_SUBSCRIPT_ELIGIBLE.contains(&[module, alias].as_slice()))
+        .map(std::string::ToString::to_string)
+        .sorted()
+        .collect();
+
+    if !names.is_empty() {
+        checker.diagnostics.push(Diagnostic::new(
+            MissingFutureAnnotationsWithImports { names },
+            stmt.range(),
+        ));
+    }
+}
+
+/// FA100
+pub fn missing_future_annotations_from_typing_usage(checker: &mut Checker, expr: &Expr) {
```

Comment:
Sorry, last question: what's the benefit in flagging both the usages and the imports? Would flagging the import alone be insufficient?

Reply:
If the user uses `import typing`, we can only flag `typing.List` at the usage. If they use `from typing import List`, then flagging the import alone is sufficient.